### ISSUE TITLE
[validation] `ExpressionError` should use resolved types.

### ISSUE THREAD

**Issue by expenses** [gfx-rs/naga#715]
At the moment, all the types used in `ExpressionError` are `Handle`s. This leads to less than ideal error messages in validation:
```
function [1] 'main' is invalid
    Expression [55] is invalid
    Operation Add can't work with [39] and [14]
```
In this case, using the `TypeInner`s would be a lot more helpful:
```
Function [1] 'main' is invalid
    Expression [55] is invalid
    Operation Add can't work with Scalar { kind: Uint, width: 4 } and Scalar { kind: Sint, width: 4 }
```

**Comment by expenses** [comment]
The root cause of this particular issue was that with a `uint` loop like: `for(uint i = 0; i < num_samples; i++)` `glslc` interprets `i++` as `i = i + int(1)` 😬

**Comment by kvark** [comment]
Thank you for filing! Having errors that are both lightweight and descriptive is hard.
wgpu-rs goes great length into resolving the handles reported by wgpu-core: https://github.com/gfx-rs/wgpu-rs/blob/master/src/backend/error.rs

**Comment by expenses** [comment]
Yeah I can definitely see the lightweight bit being a potential problem here as `TypeInner` could be a bit heavy. I was using wgpu-rs master here, and it didn't seem to resolve these handles for me. Maybe it hasn't been implemented for this expression though.

**Comment by kvark** [comment]
The wgpu-rs code I linked to resolves `wgpu-core` errors and handles, not Naga handles (yet).
The lightweight aspect is not just aesthetic, also affects stack allocation sizes, see #595
